// abp-ng2-module/esm5/lib/services/auth/token.service.js
import { __decorate } from "tslib";
///<reference path="../../../../../../node_modules/abp-web-resources/Abp/Framework/scripts/abp.d.ts"/>
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
var TokenService = /** @class */ (function () {
    function TokenService() {
    }
    TokenService.prototype.getToken = function () {
        return abp.auth.getToken();
    };
    TokenService.prototype.getTokenCookieName = function () {
        return abp.auth.tokenCookieName;
    };
    TokenService.prototype.clearToken = function () {
        abp.auth.clearToken();
    };
    TokenService.prototype.setToken = function (authToken, expireDate) {
        abp.auth.setToken(authToken, expireDate);
    };
    //refresh token
    TokenService.prototype.getRefreshToken = function () {
        return abp.auth.getRefreshToken();
    };
    TokenService.prototype.getRefreshTokenCookieName = function () {
        return abp.auth.refreshTokenCookieName;
    };
    TokenService.prototype.clearRefreshToken = function () {
        abp.auth.clearRefreshToken();
    };
    TokenService.prototype.setRefreshToken = function (refreshToken, expireDate) {
        abp.auth.setRefreshToken(refreshToken, expireDate);
    };
    TokenService.ɵprov = i0.ɵɵdefineInjectable({ factory: function TokenService_Factory() { return new TokenService(); }, token: TokenService, providedIn: "root" });
    TokenService = __decorate([
        Injectable({
            providedIn: 'root'
        })
    ], TokenService);
    return TokenService;
}());
export { TokenService };
