// abp-ng2-module/esm5/lib/services/message/message.service.js
import { __decorate } from "tslib";
///<reference path="../../../../../../node_modules/abp-web-resources/Abp/Framework/scripts/abp.d.ts"/>
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
var MessageService = /** @class */ (function () {
    function MessageService() {
    }
    MessageService.prototype.info = function (message, title, options) {
        return abp.message.info(message, title, options);
    };
    MessageService.prototype.success = function (message, title, options) {
        return abp.message.success(message, title, options);
    };
    MessageService.prototype.warn = function (message, title, options) {
        return abp.message.warn(message, title, options);
    };
    MessageService.prototype.error = function (message, title, options) {
        return abp.message.error(message, title, options);
    };
    MessageService.prototype.confirm = function (message, title, callback, options) {
        return abp.message.confirm(message, title, callback, options);
    };
    MessageService.ɵprov = i0.ɵɵdefineInjectable({ factory: function MessageService_Factory() { return new MessageService(); }, token: MessageService, providedIn: "root" });
    MessageService = __decorate([
        Injectable({
            providedIn: 'root'
        })
    ], MessageService);
    return MessageService;
}());
export { MessageService };
