// abp-ng2-module/esm2015/lib/services/message/message.service.js
import { __decorate } from "tslib";
///<reference path="../../../../../../node_modules/abp-web-resources/Abp/Framework/scripts/abp.d.ts"/>
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
let MessageService = class MessageService {
    info(message, title, options) {
        return abp.message.info(message, title, options);
    }
    success(message, title, options) {
        return abp.message.success(message, title, options);
    }
    warn(message, title, options) {
        return abp.message.warn(message, title, options);
    }
    error(message, title, options) {
        return abp.message.error(message, title, options);
    }
    confirm(message, title, callback, options) {
        return abp.message.confirm(message, title, callback, options);
    }
};
MessageService.ɵprov = i0.ɵɵdefineInjectable({ factory: function MessageService_Factory() { return new MessageService(); }, token: MessageService, providedIn: "root" });
MessageService = __decorate([
    Injectable({
        providedIn: 'root'
    })
], MessageService);
export { MessageService };
