// abp-ng2-module/esm5/lib/services/notify/notify.service.js
import { __decorate } from "tslib";
///<reference path="../../../../../../node_modules/abp-web-resources/Abp/Framework/scripts/abp.d.ts"/>
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
var NotifyService = /** @class */ (function () {
    function NotifyService() {
    }
    NotifyService.prototype.info = function (message, title, options) {
        abp.notify.info(message, title, options);
    };
    NotifyService.prototype.success = function (message, title, options) {
        abp.notify.success(message, title, options);
    };
    NotifyService.prototype.warn = function (message, title, options) {
        abp.notify.warn(message, title, options);
    };
    NotifyService.prototype.error = function (message, title, options) {
        abp.notify.error(message, title, options);
    };
    NotifyService.ɵprov = i0.ɵɵdefineInjectable({ factory: function NotifyService_Factory() { return new NotifyService(); }, token: NotifyService, providedIn: "root" });
    NotifyService = __decorate([
        Injectable({
            providedIn: 'root'
        })
    ], NotifyService);
    return NotifyService;
}());
export { NotifyService };
